### PR TITLE
Add 1M microphysics to diagnostic edmf

### PR DESCRIPTION
**Pull request by trontrytel**

### CODE REVIEW

File: src/cache/diagnostic_edmf_precomputed_quantities.jl [hunk 1]
```diff
@@ -338,9 +356,25 @@ NVTX.@annotate function set_diagnostic_edmf_precomputed_quantities_do_integral!(
...
+                    Fields.field_values(Fields.level(ᶜS_q_totʲ, i - 1))
+            end
+            if precip_model isa Microphysics1Moment
+                S_q_raiʲ_prev_level =
+                    Fields.field_values(Fields.level(ᶜS_q_raiʲ, i - 1))
+                S_q_snoʲ_prev_level =
+                    Fields.field_values(Fields.level(ᶜS_q_snoʲ, i - 1))
+                S_e_totʲ_prev_level =
+                    Fields.field_values(Fields.level(ᶜS_e_totʲ, i - 1))
+                S_prev_level = Fields.field_values(Fields.level(ᶜSᵖ, i - 1))
```
Comment (szy21) [inline src/cache/diagnostic_edmf_precomputed_quantities.jl:371]: Maybe Sᵖ_prev_level will be more clear? (I don't have strong preference, so up to you.)

File: src/cache/diagnostic_edmf_precomputed_quantities.jl [hunk 2]
```diff
@@ -338,9 +356,25 @@ NVTX.@annotate function set_diagnostic_edmf_precomputed_quantities_do_integral!(
...
+                S_snow_prev_level =
```
Comment (szy21) [inline src/cache/diagnostic_edmf_precomputed_quantities.jl:372]: Same as before, maybe Sᵖ_snow_prev_level?

File: src/cache/diagnostic_edmf_precomputed_quantities.jl [hunk 3]
```diff
@@ -401,14 +435,34 @@ NVTX.@annotate function set_diagnostic_edmf_precomputed_quantities_do_integral!(
...
+                @. S_q_totʲ_prev_level = q_tot_precipitation_sources(
+                    precip_model,
+                    thermo_params,
+                    microphys_params,
+                    dt,
+                    q_totʲ_prev_level,
+                    tsʲ_prev_level,
+                )
+            end
+            if precip_model isa Microphysics1Moment
```
Comment (szy21) [inline src/cache/diagnostic_edmf_precomputed_quantities.jl:448]:
```suggestion
            elseif precip_model isa Microphysics1Moment
```
(It doesn't really matter I think, but reduces one line:)

File: src/cache/diagnostic_edmf_precomputed_quantities.jl [hunk 4]
```diff
@@ -578,15 +640,33 @@ NVTX.@annotate function set_diagnostic_edmf_precomputed_quantities_do_integral!(
...
+                            S_q_totʲ_prev_level *
+                            e_tot_0M_precipitation_sources_helper(
+                                thermo_params,
+                                tsʲ_prev_level,
+                                Φ_prev_level,
+                            )
+                        )
+                    )
+            end
+            if precip_model isa Microphysics1Moment
```
Comment (szy21) [inline src/cache/diagnostic_edmf_precomputed_quantities.jl:661]:
```suggestion
            elseif precip_model isa Microphysics1Moment
```

File: src/cache/diagnostic_edmf_precomputed_quantities.jl [hunk 5]
```diff
@@ -706,6 +794,7 @@ NVTX.@annotate function set_diagnostic_edmf_precomputed_quantities_top_bc!(
     n = n_mass_flux_subdomains(p.atmos.turbconv_model)
     (; ᶜentrʲs, ᶜdetrʲs) = p.precomputed
     (; ᶠu³⁰, ᶠu³ʲs, ᶜuʲs, ᶠnh_pressure³ʲs) = p.precomputed
+    precip_model = p.atmos.precip_model
```
Comment (szy21) [inline src/cache/diagnostic_edmf_precomputed_quantities.jl:797]:
```suggestion
    (; precip_model) = p.atmos
```
(Just to be consistent with other parts of the code)

File: src/cache/diagnostic_edmf_precomputed_quantities.jl [hunk 6]
```diff
@@ -734,11 +823,22 @@ NVTX.@annotate function set_diagnostic_edmf_precomputed_quantities_top_bc!(
...
-            p.atmos.precip_model isa Microphysics0Moment ?
-            p.precomputed.ᶜSqₜᵖʲs.:($j) : p.scratch.ᶜtemp_scalar
-        S_q_totʲ_level = Fields.field_values(Fields.level(ᶜS_q_totʲ, i_top))
-        @. S_q_totʲ_level = 0
+        if precip_model isa Union{Microphysics0Moment, Microphysics1Moment}
+            ᶜS_q_totʲ = p.precomputed.ᶜSqₜᵖʲs.:($j)
+            S_q_totʲ_level = Fields.field_values(Fields.level(ᶜS_q_totʲ, i_top))
+            @. S_q_totʲ_level = 0
+        end
+        if precip_model isa Microphysics1Moment
```
Comment (szy21) [inline src/cache/diagnostic_edmf_precomputed_quantities.jl:831]:
```suggestion
        elseif precip_model isa Microphysics1Moment
```

**Review by szy21 — APPROVED** [review]
Thanks! I only have a few minor comments. Would you like to change `diagnostic_edmfx_aquaplanet` and `gpu_diagnostic_edmfx_aquaplanet` to use 1-moment microphysics, so we can test if it works on gpu in ci?

File: src/cache/diagnostic_edmf_precomputed_quantities.jl [hunk 7]
```diff
@@ -338,9 +356,25 @@ NVTX.@annotate function set_diagnostic_edmf_precomputed_quantities_do_integral!(
...
+            end
+            if precip_model isa Microphysics1Moment
+                S_q_raiʲ_prev_level =
+                    Fields.field_values(Fields.level(ᶜS_q_raiʲ, i - 1))
+                S_q_snoʲ_prev_level =
+                    Fields.field_values(Fields.level(ᶜS_q_snoʲ, i - 1))
+                S_e_totʲ_prev_level =
+                    Fields.field_values(Fields.level(ᶜS_e_totʲ, i - 1))
+                Sᵖ_prev_level = Fields.field_values(Fields.level(ᶜSᵖ, i - 1))
+                S_snow_prev_level =
```
Comment (szy21) [inline src/cache/diagnostic_edmf_precomputed_quantities.jl:372]:
```suggestion
                Sᵖ_snow_prev_level =
```
(Is this better?)

File: src/cache/diagnostic_edmf_precomputed_quantities.jl [hunk 8]
```diff
@@ -401,14 +435,33 @@ NVTX.@annotate function set_diagnostic_edmf_precomputed_quantities_do_integral!(
...
+                    thermo_params,
+                    microphys_params,
+                    dt,
+                    q_totʲ_prev_level,
+                    tsʲ_prev_level,
+                )
+            elseif precip_model isa Microphysics1Moment
+                compute_precipitation_sources!(
+                    Sᵖ_prev_level,
+                    S_snow_prev_level,
```
Comment (szy21) [inline src/cache/diagnostic_edmf_precomputed_quantities.jl:450]:
```suggestion
                    Sᵖ_snow_prev_level,
```

File: config/model_configs/diagnostic_edmfx_rico_box.yml [hunk 9]
```diff
@@ -8,13 +8,15 @@ implicit_diffusion: true
...
+edmfx_entr_model: "Generalized"
+edmfx_detr_model: "Generalized"
 edmfx_nh_pressure: true
 edmfx_sgs_mass_flux: true
 edmfx_sgs_diffusive_flux: true
 moist: equil
-precip_model: "0M"
+apply_limiter: false
+precip_model: "1M"
+override_τ_precip: false
```
Comment (szy21) [inline config/model_configs/diagnostic_edmfx_rico_box.yml:19]: Does `override_τ_precip` do anything when using 1-moment?

File: config/model_configs/diagnostic_edmfx_trmm_box.yml [hunk 10]
```diff
@@ -7,14 +7,14 @@ implicit_diffusion: true
...
+edmfx_entr_model: "Generalized"
+edmfx_detr_model: "Generalized"
 edmfx_nh_pressure: true
 edmfx_sgs_mass_flux: true
 edmfx_sgs_diffusive_flux: true
 moist: equil
 apply_limiter: false
-precip_model: "0M"
+precip_model: "1M"
 override_τ_precip: false
```
Comment (szy21) [inline config/model_configs/diagnostic_edmfx_trmm_box.yml:18]: same here

**Review by szy21 — APPROVED** [review]
Thank you!

**Comment by szy21** [comment]
Could you also add `output_active` here: https://github.com/CliMA/ClimaAtmos.jl/blob/e3d8c1b828f175424a02aeec0a991184cbd751ab/.buildkite/pipeline.yml#L629 in this PR? This was missed in a previous PR.

File: config/model_configs/diagnostic_edmfx_rico_box.yml (hunk 9 above)
Comment (trontrytel) [reply config/model_configs/diagnostic_edmfx_rico_box.yml:19]: No, it doesn't. This overrides the 0M precipitation formation timescale, so it is not used in 1M simulations. I'll delete it from the config.

**Pull request merged**